02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  IFileAccess,
+  MeterValueType,
+  SignedMeterValuesConfig,
+  SignedMeterValueType,
+  SystemConfig,
+} from '@citrineos/base';
+import { ILogObj, Logger } from 'tslog';
+import * as crypto from 'node:crypto';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const expectPublicKey =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    for (const meterValue of meterValues) {
+      if (expectPublicKey.length > 0 && expectPublicKey[0].value !== 'Never') {
+        for (const sampledValue of meterValue.sampledValue) {
+          const signedMeterValue = sampledValue.signedMeterValue;
+
+          if (!signedMeterValue) {
+            continue;
+          }
+
+          if (
+            signedMeterValue.publicKey &&
+            signedMeterValue.publicKey.length > 0
+          ) {
+            const incomingPublicKeyIsValid =
+              await this.isMeterValueSignatureValid(signedMeterValue);
+
+            if (
+              this._signedMeterValuesConfiguration &&
+              incomingPublicKeyIsValid
+            ) {
+              await this._chargingStationSecurityInfoRepository.readOrCreateChargingStationInfo(
```

Comment:
This is purely to make sure charging station info is created if it doesn't exist?

Asking cause we don't do anything with return value.

Reply:
yup exactly, we want it to be created if it doesn't already exist

Reply:
@thanaParis do we know how the private key the charger uses for signing is provisioned?

The spec seems to be unclear on this.

Presumably, if it follows the same flow as provisioning new PnC certs, then we would not need to store the public cert here, and would instead store it when provisioned.

Reply:
The private key will not be sent over the ocpp connection--it must end up on the charger some other way, ideally by firmware update or manual entry on the HMI.

On Citrine's side I would expect the private and public keys to be added via secure file server, so that they are encrypted at rest. They would both need to be provisioned ahead of time.

Reply:
i see! so basically on both the CSMS and charger side, there's some human intervention needed to make sure those keys end up in the right place 🤔 no way to ask the CSMS to pass the keys along, it seems :(

Reply:
@thanaParis I figured the use case `A02 - Update Charging Station Certificate by request of CSMS` might be relevant here since it details how the charger can provision itself a new pair of keys on demand:
<img width="616" alt="Screenshot 2024-08-08 at 16 42 20" src="https://github.com/user-attachments/assets/8fad2019-96e4-4264-a656-0ffb0f2acfe4">